tests/common/helpers/parallel_utils.py

```diff
@@ -1,5 +1,180 @@
+import fcntl
+import logging
+import multiprocessing
+from datetime import datetime
+from enum import Enum
+from typing import Tuple
+
 from tests.common import config_reload
+from tests.common.helpers.assertions import pytest_assert as pt_assert
+from tests.common.utilities import wait_until
+
+logger = logging.getLogger(__name__)
 
 
 def config_reload_parallel_compatible(node, results, *args, **kwargs):
     return config_reload(node, *args, **kwargs)
+
+
+def is_initial_checks_active(request):
+    parallel_state_file = request.config.getoption("--parallel_state_file")
+    try:
+        current_parallel_status = read_last_line_from_file(parallel_state_file).split(',')[1]
+        logger.debug("current parallel status: {}".format(current_parallel_status))
+        return current_parallel_status.endswith("started")
+    except FileNotFoundError:
+        logger.info("File {} not found".format(parallel_state_file))
+        return False
+
+
+def read_last_line_from_file(file):
+    with open(file, 'r') as f:
+        fcntl.flock(f, fcntl.LOCK_EX)
```

Comment:
@cyw233 For this one, do you think if it's possible that we acquire a share lock instead? Since we are reading only, acquiring an exclusive lock will slow down the read process if it's happening from other processes.

For writing, I notice we're requesting for exclusive locks anyways and we won't have it until all the subprocesses finish reading - i.e releasing a share lock so i think we would still be safe.

Reply:
Hey @auspham, a shared lock will allow multiple processes to hold a shared lock on the file and read it simultaneously. In our use case, we wanted to sequentially increment the acknowledgements, for example, LC1's process will acknowledge and increment the `setup_completed` counter to 1, then LC2's process will acknowledge and increment the `setup_completed` counter to 2. That's why I'm using exclusive lock to prevent race condition. 

Hmm..I just looked at my `acknowledge_status()` and realized that I'm doing `self._read_state()` and `self._write_state()` separately, which will still cause race condition because anything can happen when we are doing `acknowledgments += 1`. Ok let me make the `acknowledge_status()` atomic!